dlpy/model.py

```diff
@@ -896,7 +903,262 @@ def predict(self, data, text_parms=None, layer_out=None, layers=None, gpu=None,
+            is provided it will make one-step-ahead forecasts for all timestamps 
+            in the input data.(given the input data has all the columns required 
+            to make prediction.) Otherwise, it will only make one forecasted series 
+            per by-group, with the length specified by the `horizon` parameter. 
+            Default : 1 
+        layer_out : string, optional
+            Specifies the settings for an output table that includes
+            layer output values. By default, all layers are included.
+            You can filter the list with the layers parameter.
+        layers : list of strings
+            Specifies the names of the layers to include in the output
+            layers table.
+        gpu : :class:`Gpu`, optional
+            When specified, the action uses graphical processing
+            unit hardware. The simplest way to use GPU processing is
+            to specify "gpu=1". In this case, the default values of
+            other GPU parameters are used. Setting gpu=1 enables all
+            available GPU devices for use. Setting gpu=0 disables GPU
+            processing.
+        buffer_size : int, optional
+            Specifies the number of observations to score in a single
+            batch. Larger values use more memory.
+            Default: 10
+        mini_batch_buf_size : int, optional
+            Specifies the size of a buffer that is used to save input
+            data and intermediate calculations. By default, each layer
+            allocates an input buffer that is equal to the number of
+            input channels multiplied by the input feature map size
+            multiplied by the bufferSize value. You can reduce memory
+            usage by specifying a value that is smaller than the
+            bufferSize. The only disadvantage to specifying a small
+            value is that run time can increase because multiple smaller
+            matrices must be multiplied instead of a single large
+            matrix multiply.
+        use_best_weights : bool, optional
+            When set to True, the weights that provides the smallest loss
+            error saved during a previous training is used while scoring
+            input data rather than the final weights from the training.
+            default: False
+        n_threads : int, optional
+            Specifies the number of threads to use. If nothing is set then
+            all of the cores available in the machine(s) will be used.
+        casout : dict or :class:`CASTable`, optional
+            If it is dict, it specifies the output CASTable parameters.
+            If it is CASTable, it is the CASTable that will be overwritten.
+            None means a new CASTable with random name will be generated.
+            Default: None
+            
+        Returns
+        -------
+        :class:`CASTable`
+
+        """
+        
+        if not isinstance(self.train_tbl, TimeseriesTable):
```

Comment:
`self.train_tbl` and `self.valid_tbl` gets set only after calling `Model.fit`, and `self.valid_tbl` may be `None`, so may want to check this

Reply:
@yg2288 , Thanks for spotting it! In the modified merge request, I set `self.train_tbl` and `self.valid_tbl` default to be `None`. if `self.train_tbl` is still `None` when use `model.forecast` (maybe the model has not been fitted and is restored or imported from other sources), the user can manually pass a reference to the training `TimeseriesTable` through the `train_table` argument in the `model.forecast`. These information stored in the training `TimeseriesTable`  is essential for multi-step-ahead timeseries forecast, which is the main purpose of `model.forecast`. Without those information(like time interval, timeid, subsequence length etc. ), you cannot do multi-step-forecast, and should just use `model.predict` for non-timeseries prediction. I added these detailed explanations in error messages and documentations as well. 

`self.valid_tbl` can be `None`, and it won't affect down stream analysis. All the places that uses `self.valid_tbl` can handle the case where it is `None`.

Reply:
Awesome, thanks!